plugins/application/print/main.go

```diff
@@ -0,0 +1,101 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/infrawatch/apputils/logging"
+	"github.com/infrawatch/sg-core/pkg/application"
+	"github.com/infrawatch/sg-core/pkg/config"
+	"github.com/infrawatch/sg-core/pkg/data"
+)
+
+type configT struct {
+	MetricOutput string
+	EventsOutput string
+}
+
+//Print plugin suites for logging both internal buses to a file.
+type Print struct {
+	configuration configT
+	logger        *logging.Logger
+}
+
+//New constructor
+func New(logger *logging.Logger) application.Application {
+	return &Print{
+		configuration: configT{
+			MetricOutput: "/dev/stdout",
+			EventsOutput: "/dev/stdout",
+		},
+		logger: logger,
+	}
+}
+
+//Run run scrape endpoint
+func (print *Print) Run(ctx context.Context, eChan chan data.Event, mChan chan []data.Metric, done chan bool) {
+
+	metrF, err := os.OpenFile(print.configuration.MetricOutput, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
+	if err != nil {
+		print.logger.Metadata(logging.Metadata{"plugin": "print", "error": err})
+		print.logger.Error("failed to open metrics data output file")
+	} else {
+		defer metrF.Close()
+	}
+
+	evtsF, errr := os.OpenFile(print.configuration.EventsOutput, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
+	if err != nil {
+		print.logger.Metadata(logging.Metadata{"plugin": "print", "error": err})
+		print.logger.Error("failed to open events data output file")
+	} else {
+		defer evtsF.Close()
+	}
+
+	if err == nil && errr == nil {
+		print.logger.Metadata(logging.Metadata{"plugin": "print", "events": print.configuration.EventsOutput, "metrics": print.configuration.MetricOutput})
+		print.logger.Info("writing processed data to files.")
+
+		for {
+			select {
+			case <-ctx.Done():
+				goto done
+			case event := <-eChan:
+				encoded, err := json.MarshalIndent(event, "", "  ")
+				if err != nil {
+					print.logger.Metadata(logging.Metadata{"plugin": "print", "data": event})
+					print.logger.Warn("failed to marshal event data")
+				}
+				evtsF.WriteString(fmt.Sprintf("Processed event:\n%s\n", string(encoded)))
+			case metrics := <-mChan:
+				encoded, err := json.MarshalIndent(metrics, "", "  ")
+				if err != nil {
+					print.logger.Metadata(logging.Metadata{"plugin": "print", "data": metrics})
+					print.logger.Warn("failed to marshal metric data")
+				}
+				metrF.WriteString(fmt.Sprintf("Processed metric:\n%s\n", string(encoded)))
+			}
+		}
+	}
+done:
+	_, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	print.logger.Metadata(logging.Metadata{"plugin": "print"})
+	print.logger.Info("exited")
+}
+
+//Config implements application.Application
+func (print *Print) Config(c []byte) error {
+	print.configuration = configT{
```

Comment:
@paramite just a heads up, you can set defaults by initializing the structure with values. They are overwritten if a value is provided by the config file

Reply:
Theoretically this can work, but practically it does not:

2021-01-25 11:38:02 [ERROR] failed to open metrics data output file [error: open : no such file or directory, plugin: print]
2021-01-25 11:38:02 [ERROR] failed to open events data output file [plugin: print, error: open : no such file or directory]

Reply:
After revert of this commit:
2021-01-25 12:11:56 [INFO] writing processed data to files. [plugin: print, events: /dev/stdout, metrics: /dev/stdout]

Reply:
huh I'm suprised, I made that work in other instances of using this method for config files.